perf(ngcc): read dependencies from entry-point manifest

petebacondarwin opened the pull request:
Previously, even if an entry-point did not need to be processed,
ngcc would always parse the files of the entry-point to compute
its dependencies. This can take a lot of time for large node_modules.

Now these dependencies are cached in the entry-point manifest,
and read from there rather than computing them every time.

See https://github.com/angular/angular/issues/36414#issuecomment-608401834
FW-2047

petebacondarwin committed:
perf(ngcc): read dependencies from entry-point manifest

Previously, even if an entry-point did not need to be processed,
ngcc would always parse the files of the entry-point to compute
its dependencies. This can take a lot of time for large node_modules.

Now these dependencies are cached in the entry-point manifest,
and read from there rather than computing them every time.

See https://github.com/angular/angular/issues/36414\#issuecomment-608401834
FW-2047

JoostK reviewed (approved):
Nice!

gkalpak commented on packages/compiler-cli/ngcc/src/dependencies/dependency_host.ts:
```diff
@@ -21,6 +22,11 @@ export interface DependencyInfo {
 
+export interface EntryPointWithDependencies {
+  entryPoint: EntryPoint;
+  dependencies: DependencyInfo;
```
It is a little confuding that there is a `dependencies` property that has a `dependencies` property. How about renaming this to `dependencyInfo` (or `depInfo` 😱)?

petebacondarwin replied:
Sure!

gkalpak reviewed (approved):
It would be great to see some numbers. Other than that looks great ✨

gkalpak commented on packages/compiler-cli/ngcc/src/dependencies/dependency_resolver.ts:
```diff
@@ -120,18 +120,21 @@ export class DependencyResolver {
-              entryPoint.path}'.`);
+  getEntryPointWithDependencies(entryPoint: EntryPoint): EntryPointWithDependencies {
+    const dependencies = createDependencyInfo();
+    if (entryPoint.compiledByAngular) {
```
This seems like a independent fix (which I think would also fix #36524). Shouldn't it be on a separate commit?

petebacondarwin replied:
This is not actually a fix. This was already being handled in `computeDependencyGraph()` where we only add nodes to the graph if they are compiled by Angular.

Previously, this method (in the form of `getEntryPointDependencies()`) was only called on nodes in the graph, which had already been filtered by being compiled by Angular. Now this method is called on every entry-point, so we need to guard against it here too.

With that in mind, I don't think this would solve #36524. See my comment there...

petebacondarwin commented on packages/compiler-cli/ngcc/src/entry_point_finder/directory_walker_entry_point_finder.ts:
```diff
@@ -32,11 +33,11 @@ export class DirectoryWalkerEntryPointFinder implements EntryPointFinder {
       const entryPoints = this.entryPointManifest.readEntryPointsUsingManifest(basePath) ||
           this.walkBasePathForPackages(basePath);
-      unsortedEntryPoints.push(...entryPoints);
+      entryPoints.forEach(e => unsortedEntryPoints.push(e));
```
When I was looking at the generated JS code for this library, I noticed that the current code gets downleveled to ([line 43](https://unpkg.com/browse/@angular/compiler-cli@9.1.1/ngcc/src/entry_point_finder/directory_walker_entry_point_finder.js)):

```js
unsortedEntryPoints.push.apply(unsortedEntryPoints, __spread(entryPoints));
```

See [this playground](https://www.typescriptlang.org/play/?downlevelIteration=true&target=1#code/MYewdgzgLgBApmKAnAngBRAS0RGBeGAbQF0BuAWAChRJYBXSEJKOAEwFFFUNspcCSFSgwhMWHLuiw4AdAAc6EABYAKGeoTIpvCAEpSQA)

where `__spread()` is implemented like:

```js
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spread = (this && this.__spread) || function () {
    for (var ar = [], i = 0; i < arguments.length; i++) ar = ar.concat(__read(arguments[i]));
    return ar;
};
```

This seemed to me highly sub-optimal and the approach in this PR is much simpler - most likely faster and certainly creates less memory pressure.

---

This brings up a side-point.  If ngcc (and ngc) is always going to be run with node.js >= 10.0 then perhaps we should change our downleveling settings?

JoostK replied:
Yeah, that's a fair point. `__spread` usually shows up in profiling sessions quite noticeable so I would love to get rid of it. Emitting to ES2015 seems like the best way to achieve that, indeed.

petebacondarwin committed:
fixup! perf(ngcc): read dependencies from entry-point manifest

petebacondarwin commented:
Initial tests on AIO... I first built all the packages in the `node_modules` directory by running `ngcc -l debug`. Then ran the same command again a few times, which represents the "noop" ngcc run.

  | Master | PR
-- | -- | --
Full | 308.61 | 321.08
No manifest | 4.27 | 6.68
Noop 1 | 5.26 | 1.21
Noop 1 | 5.16 | 1.72
Noop 1 | 4.92 | 1.78
Noop 1 | 4.46 | 1.75
**Average noop** | **4.95** | **1.615**

So it can be seen that there is no real difference if the manifest has to be created.
But where there is a manifest, you can see that this PR is about 3x faster than `master`.

petebacondarwin committed:
perf(ngcc): reduce the size of the entry-point manifest file

The base path for package and entry-points is known so there is
no need to store these in the file. Also this commit avoids storing
empty arrays unnecessarily.

petebacondarwin committed:
perf(ngcc): only load  if it is needed

petebacondarwin commented:
Further improvements, particularly only loading the `Transformer` code if it is actually required knocks a bit more time off the noop runs:

  | Master | PR (no lazy) | PR (with lazy)
-- | -- | -- | --
noop 1 | 5.26 | 1.21 | 0.97
noop 2 | 5.16 | 1.72 | 0.96
noop 3 | 4.92 | 1.78 | 0.89
noop 4 | 4.46 | 1.75 | 0.97
**Average noop** | **4.95** | **1.615** | **0.9475**

petebacondarwin commented:
@JoostK and @gkalpak - I added a couple of commits. PTAL

gkalpak commented on packages/compiler-cli/ngcc/src/packages/entry_point_manifest.ts:
```diff
@@ -156,24 +171,29 @@ export class EntryPointManifest {
+export type EntryPointPaths = [
+  string,
+  string,
+  Array<AbsoluteFsPath>?,
+  Array<AbsoluteFsPath|PathSegment>?,
+  Array<AbsoluteFsPath>?,
```
Would it make sense to use relative paths for these too?

petebacondarwin replied:
I thought about it for `dependencies` and `deep-imports` even though there is a chance that they might start with `..` (but that probably doesn't matter). But I didn't want to do it for `missing` since some of these are `PathSegment`s but relative to the import place in the code, so we couldn't be sure, when reading in the manifest, whether we are reading in an AbsolutePath that needs de-relativising or a PathSegment that needs to be left alone.

gkalpak commented on packages/compiler-cli/ngcc/src/main.ts:
```diff
@@ -315,6 +314,7 @@ export function mainNgcc({
   const createCompileFn: CreateCompileFn = onTaskCompleted => {
     const fileWriter = getFileWriter(
         fileSystem, logger, pkgJsonUpdater, createNewEntryPointFormats, errorOnFailedEntryPoint);
+    const {Transformer} = require('./packages/transformer');
```
I assume we can do more of this. Is `Trasnformer` special in some way (e.g. particularly heavy to load)?

petebacondarwin replied:
Yes, Transformer brings in all the trait compiler stuff, all the reflection hosts etc.

petebacondarwin replied:
I looked at other imports but most of them are needed in the top level main function :-/
